schemas/authoring-guide.md

```diff
@@ -0,0 +1,829 @@
+
+* Null: may appear as a typedef'd `null` but there is discussion regarding the possibility of changing the semantics of how Null is used in Schemas. It is not commonly useful outside of the `nullable` signifier for Struct fields.
+* Boolean: may appear as `Bool` for a component specifier or `bool` as a typedef.
+* Integer: May appear as `Int` for a component specifier or `int` as a typedef. There are no additional specifiers for integer size or signedness (although this may appear as adjuncts for codegen in the future).
+* Float: May appear as `Float` for a component specifier or `float` as a typedef. There are no additional specifiers for size or byte representation (although this may appear as adjuncts for codegen in the future).
+* String: May appear as `String` for a component specifier or `string` as a typedef. The Data Model assumes unicode. Specific string encodings also appear as representation forms, see below.
+* Bytes: May appear as `Bytes` for a component specifier or `bytes` as a typedef. There are no additional specifiers for byte array length and there is no way to specify a single byte. The `byteprefix` Union representation type is a special case indicating a single byte dictates the type of the proceeding bytes, see below.
+* List: Is inferred by the `[Type]` shorthand for both typedefs and inline component specification. The token "List" is not used in the Schema DSL and all Lists must have value type specified (although Unions allow for significant flexibility here).
+* Map: Is inferred by the `{KeyType:ValueType}` shorthand for both typedefs and inline component specification. The token "Map" is not used in the Schema DSL and all Maps must have key and value type specified (although Unions allow for significant flexibility here).
+* Link: The `&` token prefixing a type is used as a shorthand for links. A generic link to an untyped resource uses the special `&Any`, while a link where there is an expected type to be found uses that type name as a hinting mechanism, `&Foo`. See below and [Links and IPLD Schemas](./links.md) for more information.
+* Union: Appears as `union` following `type` and the Union's type name.
+* Struct: Appears as `struct` following `type` and the Struct's type name.
+* Enum: Appears as `enum` following `type` and the Enum's type name.
+* Copy: Uses the shorthand `=` to indicate a copy type, as in `type Foo = Bar`. The token "Copy" does not directly appear in the Schema DSL.
+
+
+## Naming Types
+
+By convention, type names should begin with a capital letter although this requirement is not strict. Type names must only contain contain printable ASCII characters excluding space and punctuation other than underscores and the first character should be a letter. A strict regular expression for type names would be: `[a-zA-Z][a-zA-Z0-9_]*`. A regular expression following convention would be: `[A-Z][a-zA-Z0-9_]*`.
```

Comment:
👍 changing to:

> Type names must only contain alphanumeric ASCII characters and underscores, and the first character should be a letter.

We're making rules here, but I'm more than comfortable with being restrictive. schema-schema states it as:

```
## There are some additional rules that should be applied:
##   - Type names should by convention begin with a capital letter;
##   - Type names must be all printable characters (no whitespace);
##   - Type names must not contain punctuation other than underscores
##     (dashes, dots, etc.).
```

"all printable characters" is pretty expansive, but I'm happy restricting it down to just alphanumeric.

@warpfork FYI this is a spec change.

Reply:
I'm thumbs up to changing it to "alphanumeric".  Wider tolerance for fascinating characters will pretty much only ever result to more frequent headaches in codegen and etc tooling which interacts with other systems (which have their own limitations, thus potentially requiring translation rules or override supports, and so on), as far as I can tell... and that brings no one joy.

Reply:
While we're here... I've been thinking it might be nice to reserve double underscore in names as well.

In codegeneration, it often comes up that other symbols are needed which are a concatenation of the name of the type in the schema, plus some other internal munging, and having character sequences that's typically-available-in-host-languages but reserved from schema names seems like a nice simplifier.

Reply:
> begin with a capital letter

^ was intentional, btw.  If we have non-capital starter characters in golang, they'll end up needing to be capitalized, since that's semantic in golang.

And while I wouldn't generally want to take a foible of one programming language as a strong argument for anything, capitalized class names are pretty common across many languages (even if they're not always enforced), so it seems not a stretch to just formalize that requirement.

Reply:
OK, latest commit clears it up here and in schema-schema:

> Type names _must_ only contain alphanumeric ASCII characters and underscores. The first character _must_ be a capital letter. Multiple connected underscores _should_ be avoided (they should be reserved for codegen purposes)

making the dunder rule a "should"

Reply:
Sounds fine to me.